File: quarkus-mongo/src/main/java/com/demos/quarkus/service/RecordService.java
package com.demos.quarkus.service;

import com.demos.quarkus.entity.RecordDocument;
import com.demos.quarkus.repository.PrimaryRecordRepository;
import com.demos.quarkus.repository.SecondaryRecordRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.BadRequestException;
import org.bson.Document;
import org.bson.types.ObjectId;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

@ApplicationScoped
public class RecordService {

    private static final Set<String> VALID_TENANTS = Set.of("tenant1", "tenant2", "tenant3");
    private static final String[] TENANTS = {"tenant1", "tenant2", "tenant3"};
    private static final int SEED_TOTAL = 100_000;
    private static final int SEED_BATCH_SIZE = 1_000;

    @Inject
    PrimaryRecordRepository primaryRepo;

    @Inject
    SecondaryRecordRepository secondaryRepo;

    public void validateTenant(String tenant) {
        if (tenant == null || !VALID_TENANTS.contains(tenant)) {
            throw new BadRequestException("Missing or invalid X-Tenant-ID header. Allowed: tenant1, tenant2, tenant3");
        }
    }

    public Map<String, Object> createRecord(String tenant, String name, String value) {
        validateTenant(tenant);
        RecordDocument primaryRecord = new RecordDocument(tenant, name, value);
        RecordDocument secondaryRecord = new RecordDocument(tenant, name, value);

        Map<String, Object> primaryResult = primaryRepo.insert(primaryRecord);
        Map<String, Object> secondaryResult = secondaryRepo.insert(secondaryRecord);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("primary", primaryResult);
        response.put("secondary", secondaryResult);
        return response;
    }

    public Map<String, Object> getAllRecords() {
        List<Map<String, Object>> primaryRecords = primaryRepo.findAll();
        List<Map<String, Object>> secondaryRecords = secondaryRepo.findAll();

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("primary", primaryRecords);
        response.put("secondary", secondaryRecords);
        return response;
    }

    public Map<String, Object> seed() {
        long start = System.currentTimeMillis();
        int batches = SEED_TOTAL / SEED_BATCH_SIZE;

        long primaryInserted = 0;
        long secondaryInserted = 0;

        for (int batchIndex = 0; batchIndex < batches; batchIndex++) {
            int startIndex = batchIndex * SEED_BATCH_SIZE;
            List<Document> batch = createSeedBatch(startIndex, SEED_BATCH_SIZE);
            primaryRepo.insertMany(new ArrayList<>(batch));
            secondaryRepo.insertMany(createSeedBatch(startIndex, SEED_BATCH_SIZE));
            primaryInserted += batch.size();
            secondaryInserted += batch.size();
        }

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("primaryInserted", primaryInserted);
        response.put("secondaryInserted", secondaryInserted);
        response.put("durationMs", System.currentTimeMillis() - start);
        return response;
    }

    private List<Document> createSeedBatch(int startIndex, int size) {
        List<Document> batch = new ArrayList<>(size);
        LocalDateTime now = LocalDateTime.now();
        for (int i = 0; i < size; i++) {
            int globalIndex = startIndex + i;
            String tenant = TENANTS[globalIndex % TENANTS.length];
            Document doc = new Document();
            doc.append("_id", new ObjectId());
            doc.append("tenant", tenant);
            doc.append("name", "seed-name-" + globalIndex);
            doc.append("value", "seed-value-" + globalIndex);
            doc.append("createdAt", now.toString());
            batch.add(doc);
        }
        return batch;
    }
}

File: gin-mongo/internal/service/record_service.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/eliel-ontiveros/gin-mongo/internal/model"
	"github.com/eliel-ontiveros/gin-mongo/internal/repository"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

var validTenants = map[string]bool{
	"tenant1": true,
	"tenant2": true,
	"tenant3": true,
}

var tenants = []string{"tenant1", "tenant2", "tenant3"}

const (
	seedTotal     = 100_000
	seedBatchSize = 1_000
)

type RecordService struct {
	primaryRepo   *repository.PrimaryRepository
	secondaryRepo *repository.SecondaryRepository
}

func NewRecordService(primary *repository.PrimaryRepository, secondary *repository.SecondaryRepository) *RecordService {
	return &RecordService{
		primaryRepo:   primary,
		secondaryRepo: secondary,
	}
}

func (s *RecordService) ValidateTenant(tenant string) error {
	if !validTenants[tenant] {
		return fmt.Errorf("missing or invalid X-Tenant-ID header. Allowed: tenant1, tenant2, tenant3")
	}
	return nil
}

func (s *RecordService) CreateRecord(ctx context.Context, tenant, name, value string) (map[string]interface{}, error) {
	now := time.Now()

	primary := &model.RecordDocument{
		ID:        primitive.NewObjectID(),
		Tenant:    tenant,
		Name:      name,
		Value:     value,
		CreatedAt: now,
	}
	secondary := &model.RecordDocument{
		ID:        primitive.NewObjectID(),
		Tenant:    tenant,
		Name:      name,
		Value:     value,
		CreatedAt: now,
	}

	primaryResult, err := s.primaryRepo.InsertOne(ctx, primary)
	if err != nil {
		return nil, fmt.Errorf("primary insert failed: %w", err)
	}

	secondaryResult, err := s.secondaryRepo.InsertOne(ctx, secondary)
	if err != nil {
		return nil, fmt.Errorf("secondary insert failed: %w", err)
	}

	return map[string]interface{}{
		"primary":   primaryResult,
		"secondary": secondaryResult,
	}, nil
}

func (s *RecordService) GetAllRecords(ctx context.Context, limit, offset int) (map[string]interface{}, error) {
	primaryRecords, err := s.primaryRepo.FindPaginated(ctx, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("primary find failed: %w", err)
	}

	secondaryRecords, err := s.secondaryRepo.FindPaginated(ctx, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("secondary find failed: %w", err)
	}

	return map[string]interface{}{
		"primary":   primaryRecords,
		"secondary": secondaryRecords,
	}, nil
}

func (s *RecordService) Seed(ctx context.Context) (map[string]interface{}, error) {
	start := time.Now()

	primaryInserted, err := s.seedDB(ctx, true)
	if err != nil {
		return nil, fmt.Errorf("primary seed failed: %w", err)
	}

	secondaryInserted, err := s.seedDB(ctx, false)
	if err != nil {
		return nil, fmt.Errorf("secondary seed failed: %w", err)
	}

	return map[string]interface{}{
		"primaryInserted":   primaryInserted,
		"secondaryInserted": secondaryInserted,
		"durationMs":        time.Since(start).Milliseconds(),
	}, nil
}

func (s *RecordService) seedDB(ctx context.Context, primary bool) (int64, error) {
	var total int64
	batches := seedTotal / seedBatchSize

	for i := 0; i < batches; i++ {
		batch := make([]interface{}, seedBatchSize)
		for j := 0; j < seedBatchSize; j++ {
			idx := i*seedBatchSize + j
			tenant := tenants[idx%len(tenants)]
			batch[j] = &model.RecordDocument{
				ID:        primitive.NewObjectID(),
				Tenant:    tenant,
				Name:      fmt.Sprintf("seed-name-%d", idx),
				Value:     fmt.Sprintf("seed-value-%d", idx),
				CreatedAt: time.Now(),
			}
		}

		var inserted int64
		var err error
		if primary {
			inserted, err = s.primaryRepo.InsertMany(ctx, batch)
		} else {
			inserted, err = s.secondaryRepo.InsertMany(ctx, batch)
		}
		if err != nil {
			return total, err
		}
		total += inserted
	}

	return total, nil
}
